ext/socket/raddrinfo.c

```diff
@@ -369,14 +369,14 @@ rb_getaddrinfo_a(const char *node, const char *service,
 	arg.timeout = timeout;
 
 	ret = (int)(VALUE)rb_thread_call_without_gvl(nogvl_gai_suspend, &arg, RUBY_UBF_IO, 0);
-
-	if (ret) {
-	    /* on Ubuntu 18.04 (or other systems), gai_suspend(3) returns EAI_SYSTEM/ENOENT on timeout */
-	    if (ret == EAI_SYSTEM && errno == ENOENT) {
-		return EAI_AGAIN;
-	    } else {
-		return ret;
-	    }
+	if (ret && ret != EAI_ALLDONE) {
+		/* EAI_ALLDONE indicates that the request already completed and gai_suspend was redundant */
```

Comment:
This was necessary to fix tests that showed a related bug that has already existed: when resolving localhost, `gai_suspend` immediatelly returns EAI_ALLDONE because resolving localhost is fast and by the time `gai_suspend` is called, the address is already resolved.

See an example in https://gist.github.com/kirs/f1472f6c553d2d628dd359c8dd2386b7